### PR TITLE
[Feature] Add headServiceLabels field to RayCluster CR

### LINKED ISSUE #868: [Feature] Add headServiceLabels field to RayCluster CR
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

Users want to set custom labels for the KubeRay Head Service. However, currently, KubeRay only supports custom annotations for the Head Service. 

https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1673647485229799

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Yicheng-Lu-llll**
## Why are these changes needed?
Users want to set custom labels for the KubeRay Head Service. However, currently, KubeRay only supports custom annotations for the Head Service.
https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1673647485229799

## Implementation details and the reasons behind:

changes:
- In this PR,  the newly added field name is HeadServiceCustomLabels.
- In this PR, the service selector and its label are decoupled. They are no longer the same.
- In this PR, custom labels can not set keys the same as the five default labels' keys.

reasons:

1. There is a reason to call HeadServiceCustomLabels instead of HeadServiceLabels. For kubeRay head service,  `labels = five default labels + user-defined custom labels`. HeadServiceCustomLabels filed is only for user-defined custom labels. If we call HeadServiceLabels, users may think they should also add five default labels to this field. 

2. https://github.com/ray-project/kuberay/blob/d4784a5dcf9f2f1ad43a474941f0be48a2b90628/ray-operator/controllers/ray/common/service.go#L49 As shown above, in the current implementation, kubeRay head service's selector equals to its label. This fact is not true after adding custom service labels. Since the head pod will not contain these custom labels, a mismatch between the head pod and the head service will occur.

3. In the current implementation,  the service selector uses the five default labels. And head pods will have exactly the same default labels. If users overwrite any of the five default labels by setting custom service labels, a mismatch between the head pod and the head service will occur. Note: this is unlike #572, we can not change the head pod's labels.

when saying five default labels above, I am referring to
https://github.com/ray-project/kuberay/blob/d4784a5dcf9f2f1ad43a474941f0be48a2b90628/ray-operator/controllers/ray/common/service.go#L12

## What next?

In the current implementation, pod labels and service labels are assigned in different places and there is no consistency check. In other words, when modifying the code, **we need to manually make sure the value of the default five labels in the service selector and pods are the same.** 

[TODO] unit tests: whether the labels and selector match or not.

## Related issue number
Closes #868

## Checks

Step1: 
Clone my repo and checkout to `AddCustomLabels` branch

Step2: 
Follow [DEVELOPMENT.md ](https://github.com/ray-project/kuberay/blob/master/ray-operator/DEVELOPMENT.md)to build Docker image and install kuberay operator.

 Step3: 
 Add the following to the last line of kuberay/ray-operator/config/samples/ray-cluster.complete.yaml
(As @kevin85421 [mentions](https://github.com/ray-project/kuberay/pull/874#discussion_r1082149928), add  app.kubernetes.io/name: "fake_name" to test)
 ```
  headServiceCustomLabels:
    test_key0: "test_value0"
    test_key1: "test_value1"
    test_key2: "test_value2"
    app.kubernetes.io/name: "fake_name"
```

Step4: 
```
#path: kuberay/
kubectl apply -f ray-operator/config/samples/ray-cluster.complete.yaml
```

Step5: 
```
kubectl get svc
kubectl describe svc raycluster-complete-head-svc
```
ee8c91a8c6d2f1ef79a218031410a54

Step6: Set up port-forwarding
`kubectl port-forward svc/raycluster-complete-head-svc 8265:8265`

Step7: Submit a simple job
`ray job submit --address http://localhost:8265 -- python -c "import ray; ray.init(); print(ray.cluster_resources())"`

f5ea66e4e1ed78e77ad108d5c24b4c6

Step 8: In helm-chart/ray-cluster/values.yaml, add:
```
  # annotations passed on for the Head Service
  headServiceAnnotations: 
    test_ano_key0: "test_ano_value0"
  # # custom labels passed on for the Head Service
  # # to avoid conflicts, custom labels should not start with either of the prefixes `app.kubernetenes.io/` or `ray.io/`
  headServiceCustomLabels:
    test_key0: "test_value0"
    test_key1: "test_value1"
    test_key2: "test_value2"
    app.kubernetes.io/name: "fake_name"
```
Step 9
```
helm install raycluster ./helm-chart/ray-cluster
kubectl get svc
kubectl describe svc raycluster-kuberay-head-svc
```
2992b325a3f240273bc98eae137cba3

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/apis/ray/v1alpha1/raycluster_types.go [hunk 1]
```diff
@@ -21,8 +21,9 @@ type RayClusterSpec struct {
 	// EnableInTreeAutoscaling indicates whether operator should create in tree autoscaling configs
 	EnableInTreeAutoscaling *bool `json:"enableInTreeAutoscaling,omitempty"`
 	// AutoscalerOptions specifies optional configuration for the Ray autoscaler.
-	AutoscalerOptions      *AutoscalerOptions `json:"autoscalerOptions,omitempty"`
-	HeadServiceAnnotations map[string]string  `json:"headServiceAnnotations,omitempty"`
+	AutoscalerOptions       *AutoscalerOptions `json:"autoscalerOptions,omitempty"`
+	HeadServiceAnnotations  map[string]string  `json:"headServiceAnnotations,omitempty"`
+	HeadServiceCustomLabels map[string]string  `json:"HeadServiceCustomLabels,omitempty"`
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1alpha1/raycluster_types.go:26]:
```suggestion
	HeadServiceCustomLabels map[string]string  `json:"headServiceCustomLabels,omitempty"`
```

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 2]
```diff
@@ -310,6 +310,9 @@ func (r *RayClusterReconciler) reconcileServices(instance *rayiov1alpha1.RayClus
 			if val, ok := instance.Spec.HeadGroupSpec.Template.ObjectMeta.Labels[common.KubernetesApplicationNameLabelKey]; ok {
 				labels[common.KubernetesApplicationNameLabelKey] = val
 			}
+			for k, v := range instance.Spec.HeadServiceCustomLabels {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:313]: This may overwrite L311 `labels[common.KubernetesApplicationNameLabelKey] = val`. It may cause a label selector mismatch for the head Pod and head service. See https://github.com/ray-project/kuberay/pull/572 for more details.

For example, you can add a label in `HeadServiceCustomLabels` with key `common.KubernetesApplicationNameLabelKey` (i.e. "app.kubernetes.io/name") and a random value.

**Review by kevin85421 — COMMENTED** [review]
Thank you for your contribution!

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 2 above)
Comment (Yicheng-Lu-llll) [reply ray-operator/controllers/ray/raycluster_controller.go:317]: Thank you for pointing this out, I have modified it accordingly.

File: ray-operator/config/samples/ray-cluster.mini.yaml [hunk 3]
```diff
@@ -39,3 +39,4 @@ spec:
             name: dashboard
           - containerPort: 10001
             name: client
+
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-cluster.mini.yaml:42]: Remove this?

File: ray-operator/controllers/ray/common/service.go [hunk 4]
```diff
@@ -34,6 +34,12 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
 		}
 	}
 
+	selector := make(map[string]string)
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service.go:37]: What's the purpose of changes in this file?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 5]
```diff
@@ -307,6 +307,19 @@ func (r *RayClusterReconciler) reconcileServices(instance *rayiov1alpha1.RayClus
 		var err error
 		if serviceType == common.HeadService {
 			labels := make(map[string]string)
+			default_labels := map[string]struct{}{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:310]: What's the purpose of this change?

File: ray-operator/controllers/ray/common/service.go (hunk 4 above)
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service.go:37]: Merge this loop with L31 - L35.

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 5 above)
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:310]: Do not hard-code a map here. Use the function `HeadServiceLabels` instead.

**Review by kevin85421 — COMMENTED** [review]
Would you mind adding some comments about this field in RayCluster chart, `ray-cluster.complete.yaml`, and `ray-cluster.complete.large.yaml`?

**Comment by kevin85421** [comment]
Is it possible to test whether the labels and selector match or not with unit tests?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 6]
```diff
@@ -307,6 +307,13 @@ func (r *RayClusterReconciler) reconcileServices(instance *rayiov1alpha1.RayClus
 		var err error
 		if serviceType == common.HeadService {
 			labels := make(map[string]string)
+			default_labels := common.HeadServiceLabels(*instance)
+			for k, v := range instance.Spec.HeadServiceCustomLabels {
+				// CustomLabels should not change default labels
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:312]:
```suggestion
				// Custom labels should not change default labels
```

File: ray-operator/controllers/ray/common/service_test.go [hunk 7]
```diff
@@ -126,3 +126,20 @@ func TestBuildServiceForHeadPodWithAnnotations(t *testing.T) {
...
+	assert.Nil(t, err)
+
+	// Selector should not contain any custom label.
+	if _, ok := svc.Spec.Selector["key"]; ok {
+		t.Fatalf("Expected `%v` not to exist", "key")
+	}
+
+	if _, ok := svc.ObjectMeta.Labels["key"]; !ok {
+		t.Fatalf("Expected `%v`", "key")
+	}
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service_test.go:144]: We can also verify the length of `svc.Spec.Selector` and `svc.ObjectMeta.Labels` here.

**Review by kevin85421 — COMMENTED** [review]
We need to have a clear story about labels for the Head service.

# Without custom labels for Head service
* Head Pod: labels (`head_pod_labels`)
* Head Service: 
  * selector: `default_labels` (If `KubernetesApplicationNameLabelKey` is defined by `head_pod_labels`, overwrite the default value.)
  * labels: `default_labels` (If `KubernetesApplicationNameLabelKey` is defined by `head_pod_labels`, overwrite the default value.)
https://github.com/ray-project/kuberay/blob/2600854c61673f2b7da9fe2b54c8220468c1a013/ray-operator/controllers/ray/raycluster_controller.go#L309-L312

# Current status of this PR
* Head Pod: labels (`head_pod_labels`)
* Head Service: 
  * selector: `default_labels` (If `KubernetesApplicationNameLabelKey` is defined by `head_pod_labels`, overwrite the default value.)
  * labels: `default_labels` (If `KubernetesApplicationNameLabelKey` is defined by `head_pod_labels`, overwrite the default value.) + `custom_svc_labels`

# Need to discuss
* Q: Does `HeadService.labels` need to contain `default_labels`?
  * For me, it seems unnecessary. 
* Q: How to handle overlaps between `head_pod_labels`, `default_labels`, and `custom_svc_labels`?
  * In my opinion, `default_labels` should always respect `head_pod_labels` and ignore `custom_svc_labels` to avoid mismatch between label and selector.

# My current thought
* Head Pod: labels (`head_pod_labels`)
* Head Service: 
  * selector: `default_labels` (if overlaps exist, respect `head_pod_labels` and ignore `custom_svc_labels`.)
  * labels: `custom_svc_labels`

cc @DmitriGekhtman @architkulkarni @Jeffwan for more inputs.

File: ray-operator/config/samples/ray-cluster.complete.yaml [hunk 8]
```diff
@@ -133,3 +133,13 @@ spec:
           - name: ray-logs
             emptyDir: {}
 ######################status#################################
+  # annotations passed on for the Head Service
+  headServiceAnnotations: {}
+  # custom labels passed on for the Head Service
+  # custom labels should not overwrite
```
Comment (DmitriGekhtman) [inline ray-operator/config/samples/ray-cluster.complete.yaml:139]:
```suggestion
  # to avoid conflicts, custom labels should not start with either of the prefixes `app.kubernetenes.io/` or `ray.io/`
```

Comment (DmitriGekhtman) [reply ray-operator/config/samples/ray-cluster.complete.yaml:139]: (Because the particular labels set by the operator and kubernetes might change in the future but the prefixes probably won't.)

Comment (Yicheng-Lu-llll) [reply ray-operator/config/samples/ray-cluster.complete.yaml:139]: Thank you for pointing this out! I will change it accordingly.

**Comment by Jeffwan** [comment]
I feel we should not update CRD for this kind of change. Let's change to use annotations instead?? This won't be scalable and will constantly introduce more api changes. WDYT?

**Review by Jeffwan — CHANGES_REQUESTED** [review]
Let' have a short discussion on future api changes?

**Comment by DmitriGekhtman** [comment]
> I feel we should not update CRD for this kind of change. Let's change to use annotations instead?? This won't be scalable and will constantly introduce more api changes. WDYT?

The request in this case was specifically to expose head service labels (we didn't ask what for)

**Comment by kevin85421** [comment]
@Jeffwan

I followed up with the user on slack. In his words, "just FYI I'm not actively working on this, but the original need was due to use of labels for cross-cluster service discovery". 

> Let's change to use annotations instead??

It seems annotations will not work in this case.
"Labels can be used to select objects and to find collections of objects that satisfy certain conditions. In contrast, annotations are not used to identify and select objects." ([Kubernetes official doc](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#attaching-metadata-to-objects))

> This won't be scalable and will constantly introduce more api changes. WDYT?

* We have already exposed custom annotations for the Head service (https://github.com/ray-project/kuberay/pull/841).
* I think we should expose custom labels for the Head service due to the following reasons:
  * Keep consistency with the custom annotations for the Head service.
  * We cannot directly copy the labels to the Head service, and we will check the labels and attach new labels for the Head service. You can see the discussion about label overlaps above.
* We plan to expose the entire Head service except for some fields, including selector, annotations, labels, ... etc.
  *  https://github.com/ray-project/kuberay/issues/877#issuecomment-1398627415 
  *  With this, we do not need to introduce any CRD update for the Head service.

**Comment by Jeffwan** [comment]
@kevin85421 @Yicheng-Lu-llll sorry for late response. I am kind of overwhelmed by internal stuff. 

For this api change, I feel we can make it with next API version like v1beta1. I will double check your notes and come back to you soon

**Comment by kevin85421** [comment]
@Jeffwan no worries. We may close this PR in favor of exposing the entire service (https://github.com/ray-project/kuberay/issues/877#issuecomment-1398627415) based on community sync on Jan. 31 2023.
